KIPs/kip-149.md

```diff
@@ -0,0 +1,292 @@
+    event Registered(string name, address indexed addr, uint256 indexed activation);
+
+    /* ========== MUTATORS ========== */
+    /// @dev Registers a new system contract.
+    function register(string memory name, address addr, uint256 activation) external virtual;
+    
+    /// @dev Transfers ownership to newOwner.
+    function transferOwnership(address newOwner) external virtual;
+
+    /* ========== GETTERS ========== */
+    /// @dev Returns an address for active system contracts registered as name if exists.
+    ///  It returns a zero address when there's no active system contract with name.
+    function getActiveAddr(string memory name) external virtual returns (address);
+  
+    /// @dev Returns all system contracts registered as name.
+    function getAllRecords(string memory name) external virtual view returns (Record[] memory);
+    
+    /// @dev Returns all names of registered system contracts.
+    function getAllNames() external virtual view returns (string[] memory);
+    
+    /// @dev Returns owner of contract.
+    function owner() external virtual view returns (address);
+}
+```
+
+#### Methods
+
+```solidity
+function register(string memory name, address addr, uint256 activation)
+```
+
+Registers a new system contract. It will be activated at `activation`. It overwrites the predecessor if a predecessor system contract exists and is not yet active. Passing `addr == address(0)` is an implicit deprecation for the `name`, meaning the `name` will no longer be used.
+
+The function validates the following requirements:
+
+- The function caller MUST be a governance contract.
+
+- The function MUST revert if a `name` is an empty string.
+
+- The function MUST revert if `activation < block.number`.
+
+The function emits a `Registered` event.
+
+```solidity
+function getActiveAddr(string memory name) view returns (address)
+```
+
+Returns the address of the active system contract with the `name`. It returns a zero address if there’s no registered system contract with the `name` or the `name` has been deprecated by registering a zero address.
+
+#### 2. Proxy
+
+The implementation of the proxy contract will come from [OZ's UUPS implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.9.3/contracts/proxy/ERC1967/ERC1967Proxy.sol), which follows [EIP-1967](https://eips.ethereum.org/EIPS/eip-1967) and [EIP-1822](https://eips.ethereum.org/EIPS/eip-1822). 
+
+### System Contracts Life Cycle
+
+The Registry contract manages system contracts based on the current block number. Its state will be managed implicitly, which means there’s no explicit state variable(e.g., enum State). It has three implicit states and cannot be reversed to the previous state:
+
+- Registered: It has been registered but has not been activated yet.
+
+- Active: The current block number exceeds its activation, and no active successor system contract exists.
+
+- Deprecated: There’s an active successor system contract.
```

Comment:
What about registering '0'?

Reply:
Does it mean to register `activation = 0`? It might cause confusion since activating from `block=0` makes intuitive sense(and `KIP113` will be `activation = 0`). And we need additional logic to filter out `activation == 0` to search in a view function.

Anyway, I need to add a condition with zero address(deprecated name) for `Active`, and `Deprecated`.